tests/test_build_latex.py

```diff
@@ -1708,3 +1708,30 @@ def test_copy_images(app, status, warning):
         'rimg.png',
         'testimäge.png',
     }
+
+
+@pytest.mark.sphinx('latex', testroot='latex-labels-before-module')
+def test_duplicated_labels_before_module(app, status, warning):
+    app.build()
+    content: str = (app.outdir / 'python.tex').read_text()
+
+    def count_label(name):
+        text = r'\phantomsection\label{\detokenize{%s}}' % name
+        return content.count(text)
+
+    pattern = r'\\phantomsection\\label\{\\detokenize\{index:label-(?:auto-)?\d+[a-z]*}}'
+    expect_labels = {match.group() for match in re.finditer(pattern, content)}
+    result_labels = set()
+
+    # iterate over the (explicit) labels in the corresponding index.rst
+    for rst_label_name in {
+        'label_1a', 'label_1b', 'label_2', 'label_3',
+        'label_auto_1a', 'label_auto_1b', 'label_auto_2', 'label_auto_3',
+    }:
+        tex_label_name = 'index:' + rst_label_name.replace('_', '-')
+        tex_label_code = r'\phantomsection\label{\detokenize{%s}}' % tex_label_name
+        assert content.count(tex_label_code) == 1, f'duplicated label: {tex_label_name!r}'
+        result_labels.add(tex_label_code)
+
+    # sort the labels for a better visual diff, if any
+    assert sorted(result_labels) == sorted(expect_labels)
```

Comment:
In my usage of tox which is not quite same as our CI, this does not help much. Why not simply test  with len()? As the for loop succeeded we know our labels are there without duplicate, so only thing possibly wrong is if mysteriously there are other labels in output matching the regex.

Reply:
Actually, that's what I wanted to do, namely checking that all labels have been correctly checked. 

The reason why I used `sorted` is to have a diff which tells me which value is in the LHS but not in the RHS (or vice-versa). But this *may* actually be a feature for PyCharm only. I'll change that into a `len` check then.

Reply:
I confirmed that this is *NOT* a PyCharm feature but a tox feature.

__**With `len` checks**__

```text
>       assert len(tested_labels) == len(output_labels)
E       AssertionError: assert 7 == 8
E        +  where 7 = len({'\\phantomsection\\label{\\detokenize{index:label-1a}}', '\\phantomsection\\label{\\detokenize{index:label-1b}}', '\\phantomsection\\label{\\detokenize{index:label-2}}', '\\phantomsection\\label{\\detokenize{index:label-auto-1a}}', '\\phantomsection\\label{\\detokenize{index:label-auto-1b}}', '\\phantomsection\\label{\\detokenize{index:label-auto-2}}', ...})
E        +  and   8 = len(frozenset({'\\phantomsection\\label{\\detokenize{index:label-1a}}', '\\phantomsection\\label{\\detokenize{index:label-1b}}', '\\phantomsection\\label{\\detokenize{index:label-2}}', '\\phantomsection\\label{\\detokenize{index:label-3}}', '\\phantomsection\\label{\\detokenize{index:label-auto-1a}}', '\\phantomsection\\label{\\detokenize{index:label-auto-1b}}', ...}))

tests/test_build_latex.py:1738: AssertionError
```

__**With `sorted` checks**__

```txt

E         Full diff:
E           [
E            '\\phantomsection\\label{\\detokenize{index:label-1a}}',
E            '\\phantomsection\\label{\\detokenize{index:label-1b}}',
E            '\\phantomsection\\label{\\detokenize{index:label-2}}',
E         -  '\\phantomsection\\label{\\detokenize{index:label-3}}',
E            '\\phantomsection\\label{\\detokenize{index:label-auto-1a}}',
E            '\\phantomsection\\label{\\detokenize{index:label-auto-1b}}',
E            '\\phantomsection\\label{\\detokenize{index:label-auto-2}}',
E            '\\phantomsection\\label{\\detokenize{index:label-auto-3}}',
E           ]
```

Reply:
Indeed you are right, I badly tested your test here.  And to reproduce the output I had to modify tox.ini to use pytest -vv.

This final assert tests a rather exotic situation which is not the original one about duplicated labels; as a rule many tests are rather minimal in our test suite, and this one really goes to some length to cover all possibilities ;-) ok let's go with it.